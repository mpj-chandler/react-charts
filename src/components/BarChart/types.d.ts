import Placement from '../../enums/Placement';

export interface SeriesDataPoint {
    x: number | null;
    y: number | null;
}

export interface NonNullSeriesDataPoint extends SeriesDataPoint {
    x: number;
    y: number;
}

export interface SeriesData {
    seriesName: string;
    points: SeriesDataPoint[];
}

export interface Padding {
    left: number;
    right: number;
    top: number;
    bottom: number;
}

export interface AxisRange {
    min: number;
    max: number;
}

export interface AxisProps {
    data: SeriesData[];
    title: string;
    padding: Padding;
    config?: AxisConfig;
}

export interface AxisConfig {
    zeroIntercept: boolean;
    margin: number;
    tickPlacement: Placement;
    tickLength: number;
}
