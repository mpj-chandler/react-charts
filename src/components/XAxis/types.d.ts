import { AxisConfig, Padding, SeriesData } from '../BarChart/types';

export interface XAxisTickProps {
    data: SeriesData[];
    padding: Padding;
    config?: AxisConfig;
}
