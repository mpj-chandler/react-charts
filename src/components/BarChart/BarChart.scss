.BarChart {
  height: 100%;
  width: 100%;
  display: grid;
  grid-template-areas: "title title" "y-axis plot" "blank x-axis";
  grid-template-columns: 20% 60% 20%;
  grid-template-rows: 10% 70% 20%;

  &__title {
    grid-area: title;
  }

  &__blank {
    grid-area: blank;
    z-index: 0;
  }
}
