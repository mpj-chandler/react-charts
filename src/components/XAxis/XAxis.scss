.XAxis {
  grid-area: x-axis;
  z-index: 100;

  &__Svg {
    width: 100%;
    height: 100%;
  }

  &__Ticks {
    width: 100%;
    height: 100%;
  }

  &__TickLabels {
    font-size: calc(8px + 0.5vh);
  }
}
