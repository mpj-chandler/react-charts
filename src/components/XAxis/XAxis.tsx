import React from 'react';
import styles from './XAxis.scss';
import { getXAxisRange } from '../../utils/axisUtils/getXAxisRange/getXAxisRange';
import { getYAxisRange } from '../../utils/axisUtils/getYAxisRange/getYAxisRange';
import Placement from '../../enums/Placement';
import AxisTickLabel from '../AxisTickLabel/AxisTickLabel';
import Axis from '../../enums/Axis';
import generateAxisLabels from '../../utils/axisUtils/generateAxisLabels/generateAxisLabels';
import { AxisTickProps, AxisProps } from '../../__types__/axisTypes';
import { SeriesDataPoint } from '../../__types__/seriesTypes';

function getTickXPosition(props: AxisTickProps, label: number) {
    const range = getXAxisRange(props.data, props.config);
    const numPoints = props.data[0].points.length;
    const interval = range.max - range.min;
    const totalHorizontalPadding = props.padding.right + props.padding.left;

    if (props.config && props.config.tickPlacement === Placement.Bucket) {
        const step = interval / numPoints;

        return ((100 - totalHorizontalPadding) * ((step + label - range.min) / interval)) + props.padding.left;
    }

    return ((100 - totalHorizontalPadding) * (label - range.min) / interval) + props.padding.left;
}

function getTickTextXPosition(x1: number, props: AxisTickProps, label: number) {
    const range = getXAxisRange(props.data, props.config);
    const numPoints = props.data[0].points.length;
    const interval = range.max - range.min;
    const totalHorizontalPadding = props.padding.right + props.padding.left;

    if (props.config && props.config.tickPlacement === Placement.Bucket) {
        const step = 0.5 * (interval / numPoints);

        return (100 - totalHorizontalPadding) * ((step + label - range.min) / interval) + props.padding.left;
    }

    return x1;
}

function renderTick(
    label: number,
    x1: number,
    y1: number,
    props: AxisTickProps,
    textXPos: number,
    tickLength: number = 5) {
    const y2 = y1 === props.padding.top ? y1 - tickLength / 2 : y1 + tickLength;

    return (
        <g key={`tickGroup-${label}`}>
            <line
                key={`tickLine-${label}`}
                x1={`${x1}%`}
                y1={`${y1}%`}
                x2={`${x1}%`}
                y2={`${y2}%`}
                stroke={'black'}
            />
            <AxisTickLabel
                key={`tickText-${label}`}
                xPos={textXPos}
                yPos={y2}
                label={label}
                axis={Axis.XAxis}
            />
        </g>
    );
}

function renderTicks(props: AxisTickProps, y1: number) {
    const points: SeriesDataPoint[] = props.data[0].points;
    const tickLabels: string[] = generateAxisLabels(props.data, Axis.XAxis);

    return (
        <g className={styles.XAxis__Ticks}>
            {tickLabels.map((label: string) => {
                const x1 = getTickXPosition(props, Number(label));
                const textXPos = getTickTextXPosition(x1, props, Number(label));

                return renderTick(Number(label), x1, y1, props, textXPos);
            })}
        </g>
    );
}

function getXAxisYPos(props: AxisProps): number {
    const range = getYAxisRange(props.data, props.config);
    const interval = range.max - range.min;
    const totalVerticalPadding = props.padding.top + props.padding.bottom;

    if (range.max <= 0) {
        return props.padding.top;
    }

    return ((100 - totalVerticalPadding) * (range.max / interval)) + props.padding.top;
}

const XAxis: React.FC<AxisProps> = (props) => {
    const axisYPos = getXAxisYPos(props);

    return (
        <div className={styles.XAxis}>
            <svg className={styles.XAxis__Svg}>
                <line x1={`${props.padding.left}%`} y1={`${axisYPos}%`} x2={`${100 - props.padding.right / 2}%`} y2={`${axisYPos}%`} stroke={'black'} fill={'transparent'} strokeWidth={1}/>
                {renderTicks({data: props.data, padding: props.padding, config: props.config}, axisYPos)}
            </svg>
        </div>
    );
};

export default XAxis;
