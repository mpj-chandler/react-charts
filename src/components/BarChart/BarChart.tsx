import React from 'react';
import styles from './BarChart.scss';
import XAxis from '../XAxis/XAxis';
import YAxis from '../YAxis/YAxis';
import { SeriesData, Padding } from './types';

export interface BarChartProps {
    data: SeriesData[];
    title: string;
    padding: Padding;
}

const BarChart: React.FC<BarChartProps> = (props: BarChartProps) => {
    return (
        <div className={styles.BarChart}>
            <div className={styles.BarChart__title}>{props.title}</div>
            <YAxis title={'YAxis'} data={props.data} padding={props.padding}/>
            <div className={styles.BarChart__blank}>()</div>
            <XAxis title={'XAxis'} data={props.data} padding={props.padding}/>
            <div className={styles.BarChart__padRight}>()</div>
        </div>
    )
}

export default BarChart;
