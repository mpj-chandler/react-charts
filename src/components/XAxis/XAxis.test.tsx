import React from 'react';
import TestRenderer from 'react-test-renderer';
import XAxis from './XAxis';
import { AxisProps } from '../BarChart/types';

describe('XAxis', () => {
    const props: AxisProps = {
        title: 'anything',
        data: [
            {
                seriesName: 'A',
                points: [
                    {
                        x: 0,
                        y: 10
                    },
                    {
                        x: 1,
                        y: 20
                    },
                    {
                        x: 2,
                        y: 30
                    }
                ]
            }
        ],
        padding: {
            top: 10,
            bottom: 10,
            left: 10,
            right: 10
        }
    };

    it('it renders consistently', () => {
        const component = TestRenderer.create(<XAxis {...props}/>);

        expect(component.toJSON()).toMatchSnapshot();
    })
})