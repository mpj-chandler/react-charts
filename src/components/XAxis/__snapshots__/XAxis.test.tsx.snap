// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`XAxis it renders consistently 1`] = `
<div
  className="XAxis"
>
  <svg
    className="XAxis__Svg"
  >
    <line
      fill="transparent"
      stroke="black"
      strokeWidth={1}
      x1="0%"
      x2="100%"
      y1="0%"
      y2="0%"
    />
    <g
      className="XAxis__Ticks"
    >
      <g>
        <line
          stroke="black"
          x1="0%"
          x2="0%"
          y1="0%"
          y2="10%"
        />
        <text
          className="XAxis__TickLabels"
          stroke="black"
          x="0%"
          y="25%"
        >
          0
        </text>
      </g>
      <g>
        <line
          stroke="black"
          x1="45%"
          x2="45%"
          y1="0%"
          y2="10%"
        />
        <text
          className="XAxis__TickLabels"
          stroke="black"
          x="45%"
          y="25%"
        >
          1
        </text>
      </g>
      <g>
        <line
          stroke="black"
          x1="90%"
          x2="90%"
          y1="0%"
          y2="10%"
        />
        <text
          className="XAxis__TickLabels"
          stroke="black"
          x="90%"
          y="25%"
        >
          2
        </text>
      </g>
    </g>
  </svg>
</div>
`;
